src/uct/tcp/tcp_cm.c

```diff
@@ -85,190 +92,405 @@ static void uct_tcp_cm_trace_conn_pkt(const uct_tcp_ep_t *ep, const char *msg,
-            return 0;
+        return status;
+    }
+
+    if ((cmp > 0) ||
+        (connect_ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED)) {
+        /* Have to live with the current connection and reject this one */
+
+        /* 1. Migrate RX from the EP allocated during accepting connection to
+         *    the found EP. Don't set anything if != CONNECTED, because we need
+         *    to handle a connection data */
+        status = uct_tcp_ep_move_ctx_cap(accept_ep, connect_ep,
+                                         UCT_TCP_EP_CTX_TYPE_RX);
+        if (status != UCS_OK) {
+            return status;
         }
-        goto err;
+
+        if (connect_ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED) {
+            uct_tcp_ep_mod_events(connect_ep, EPOLLIN, 0);
+        }
+
+        /* 2. Destroy the EP allocated during accepting connection */
+        uct_tcp_ep_destroy_internal(&accept_ep->super.super);
+    } else /* our iface address less than remote && we are not connected */ {
+        /* Have to accept this connection and close the current one */
+
+        /* 1. Accept the remote connection and close the current one */
+        status = uct_tcp_cm_simult_conn_accept_remote_conn(accept_ep, connect_ep,
+                                                           progress_count);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        /* 2. Progress TX, because we may have some pending operations queued */
+        *progress_count += uct_tcp_ep_progress(connect_ep,
+                                               UCT_TCP_EP_CTX_TYPE_TX);
     }
 
-    status = uct_tcp_cm_send_conn_req(ep);
+    return status;
+}
+
+static unsigned
+uct_tcp_cm_handle_conn_req(uct_tcp_ep_t **ep_p,
+                           const uct_tcp_cm_conn_req_pkt_t *cm_req_pkt)
+{
+    uct_tcp_ep_t *ep        = *ep_p;
+    uct_tcp_iface_t *iface  = ucs_derived_of(ep->super.super.iface,
+                                             uct_tcp_iface_t);
+    unsigned progress_count = 0;
+    ucs_status_t status;
+    uct_tcp_ep_t *pair_ep;
+
+    ep->peer_addr = cm_req_pkt->iface_addr;
+    
+    status = uct_tcp_ep_add_ctx_cap(ep, UCT_TCP_EP_CTX_TYPE_RX);
     if (status != UCS_OK) {
-        return 0;
+        goto err;
     }
 
-    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_WAITING_ACK);
-    uct_tcp_ep_mod_events(ep, EPOLLIN, EPOLLOUT);
+    if (ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED) {
+        status = uct_tcp_cm_send_conn_ack(ep);
+        if (status != UCS_OK) {
+            goto err;
+        }
+        return 1;
+    }
 
-    ucs_assertv((ep->tx.length == 0) && (ep->tx.offset == 0) &&
-                (ep->tx.buf == NULL), "ep=%p", ep);
-    return 1;
+    ucs_assertv(!(ep->ctx_caps & UCS_BIT(UCT_TCP_EP_CTX_TYPE_TX)),
+                "ep %p mustn't have TX cap", ep);
+
+    if (!uct_tcp_ep_peer_addr_to_itself(ep) &&
+        (pair_ep = uct_tcp_cm_search_ep(iface, &ep->peer_addr,
```

Comment:
what if it already has RX and TX types?
What is the general logic? Is it like below?
- We create an ep and there is only RX-capable ep, we try to reuse it
- We create and ep and there is TX/RX ep, we create a new ep
- We create an ep and there is only TX-capable ep, we create a new ep
- We receive CONN_REQ and there is only TX-cap ep - we try to reuse it
- We receive CONN_REQ and there is TX/RX ep - we create a new "priv RX" ep 
right?

Reply:
> * We receive CONN_REQ and there is only TX-cap ep - we try to reuse it

this is only the case when we receive CM CONN_REQ event from the peer, because CONN_REQ can be sent by remote side only in case of:
1. it connects to us - we have already accepted the connection (`*ep_p` is the EP created from accepting)
2. we have already been connected (we were an initiator of this connection) to this remote side some time ago, now it just wants to send the data to us - so, `*ep_p` has only TX cap and after handling of this CONN_REQ, we will have TX|RX caps.